free/src/main/scala/cats/free/Free.scala

```diff
@@ -130,31 +88,141 @@ sealed abstract class Free[S[_], A] extends Product with Serializable {
       }
     }(Free.freeMonad)
 
-  final def compile[T[_]](f: FunctionK[S,T]): Free[T, A] = mapSuspension(f)
+  override def toString(): String =
+    "Free(...)"
+}
+
+object Free {
+
+  /**
+   * Return from the computation with the given value.
+   */
+  private[free] final case class Pure[S[_], A](a: A) extends Free[S, A]
+
+  /** Suspend the computation with the given suspension. */
+  private[free] final case class Suspend[S[_], A](a: S[A]) extends Free[S, A]
+
+  /** Call a subroutine and continue with the given function. */
+  private[free] final case class Gosub[S[_], B, C](c: Free[S, C], f: C => Free[S, B]) extends Free[S, B]
+
+  /**
+   * Lift a pure `A` value into the free monad.
+   */
+  def pure[S[_], A](a: A): Free[S, A] = Pure(a)
+
+  /**
+   * Lift an `F[A]` value into the free monad.
+   */
+  def liftF[F[_], A](value: F[A]): Free[F, A] = Suspend(value)
+
+  /**
+   * Suspend the creation of a `Free[F, A]` value.
+   */
+  def suspend[F[_], A](value: => Free[F, A]): Free[F, A] =
+    pure(()).flatMap(_ => value)
 
-  override def toString(): String = "Free(...)"
+  /**
+   * This method is used to defer the application of an Inject[F, G]
+   * instance. The actual work happens in
+   * `FreeInjectPartiallyApplied#apply`.
+   *
+   * This method exists to allow the `F` and `G` parameters to be
+   * bound independently of the `A` parameter below.
+   */
+  def inject[F[_], G[_]]: FreeInjectPartiallyApplied[F, G] =
+    new FreeInjectPartiallyApplied
+
+  /**
+   * Pre-application of an injection to a `F[A]` value.
+   */
+  final class FreeInjectPartiallyApplied[F[_], G[_]] private[free] {
+    def apply[A](fa: F[A])(implicit I: Inject[F, G]): Free[G, A] =
+      Free.liftF(I.inj(fa))
+  }
+
+  /**
+   * `Free[S, ?]` has a monad for any type constructor `S[_]`.
+   */
+  implicit def freeMonad[S[_]]: MonadRec[Free[S, ?]] =
+    new MonadRec[Free[S, ?]] {
+      def pure[A](a: A): Free[S, A] = Free.pure(a)
+      override def map[A, B](fa: Free[S, A])(f: A => B): Free[S, B] = fa.map(f)
+      def flatMap[A, B](a: Free[S, A])(f: A => Free[S, B]): Free[S, B] = a.flatMap(f)
+      def tailRecM[A, B](a: A)(f: A => Free[S, A Xor B]): Free[S, B] =
+        f(a).flatMap(_ match {
+          case Left(a1) => tailRecM(a1)(f) // recursion OK here, since Free is lazy
+          case Right(b) => pure(b)
+        })
+    }
+
+  /**
+   * Perform a stack-safe monadic fold from the source context `F`
+   * into the target monad `G`.
+   *
+   * This method can express short-circuiting semantics. Even when
+   * `fa` is an infinite structure, this method can potentially
+   * terminate if the `foldRight` implementation for `F` and the
+   * `tailRecM` implementation for `G` are sufficiently lazy.
+   */
+  def foldLeftM[F[_]: Foldable, G[_]: MonadRec, A, B](fa: F[A], z: B)(f: (B, A) => G[B]): G[B] =
+    unsafeFoldLeftM[F, Free[G, ?], A, B](fa, z) { (b, a) =>
+      Free.liftF(f(b, a))
+    }.runTailRec
```

Comment:
@TomasMikula as the author of this code, I'll be the first to say that it is a bit bizarre. `G.tailRecM` should be being called by the `runTailRec` call on the `Free` structure. `Eval` makes an appearance due to the `foldRight` type signature, but really we are just using `foldRight` because of its laziness, and `unsafeFoldLeftM` is calling `.value` on the `Eval`, so we don't actually get any `Eval` trampolining here. It's effectively like we are using a scalaz-style `foldRight` that takes a byname parameter, since we aren't taking advantage `Eval`'s trampolining.

If you can figure out a way to provide a more direct implementation, I'd definitely be interested. I think that this has to be done in terms of `foldRight` if we want laziness, though, which means `Eval` will definitely be in play. You may be able to use `MonadRec` and `foldRight` without `Free`, and if so that'd be great - as it stands I'm kind of convinced that we should move `Free` (the monad but not the other free structures) back into core for this sort of thing.